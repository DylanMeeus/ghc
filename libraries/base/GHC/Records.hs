{-# LANGUAGE NoImplicitPrelude
           , MultiParamTypeClasses
           , MagicHash
           , KindSignatures
           , DataKinds
           , FunctionalDependencies
  #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  GHC.Records
-- Copyright   :  (c) Adam Gundry 2015
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  cvs-ghc@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC extensions)
--
--
-----------------------------------------------------------------------------

module GHC.Records
       ( HasField(..)
       ) where

import GHC.Base ( Symbol )
import GHC.Exts ( Proxy# )

class HasField (x :: Symbol) r a | x r -> a where
  getField :: Proxy# x -> r -> a
