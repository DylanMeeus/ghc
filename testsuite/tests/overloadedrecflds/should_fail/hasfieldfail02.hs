{-# LANGUAGE DataKinds, ExistentialQuantification, MagicHash, RankNTypes #-}

import GHC.Prim (Proxy#, proxy#)
import GHC.Records (HasField(..))

data T = MkT { foo :: forall a . a -> a }
data U = forall b . MkU { bar :: b }

-- This should fail because foo is higher-rank.
x = getField (proxy# :: Proxy# "foo") (MkT id)

-- This should fail because bar is a naughty record selector (it
-- involves an existential).
y = getField (proxy# :: Proxy# "bar") (MkU True)

main = return ()
