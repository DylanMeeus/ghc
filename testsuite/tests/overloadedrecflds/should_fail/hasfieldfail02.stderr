
hasfieldfail02.hs:10:5: error:
    • No instance for (HasField "foo" T a1)
        arising from a use of ‘getField’
    • In the expression: getField (proxy# :: Proxy# "foo") (MkT id)
      In an equation for ‘x’:
          x = getField (proxy# :: Proxy# "foo") (MkT id)

hasfieldfail02.hs:14:5: error:
    • No instance for (HasField "bar" U a0)
        arising from a use of ‘getField’
    • In the expression: getField (proxy# :: Proxy# "bar") (MkU True)
      In an equation for ‘y’:
          y = getField (proxy# :: Proxy# "bar") (MkU True)
